### PR TITLE
Revamp ToC ticks handling

### LINKED ISSUE #5500: When chapters have subchapters
#### Issue
When a chapter has subchapters the Mini progress bar has chapter markers only for the subchapters, but not for the chapter itself. 
For example in the book below there is a marker for 'Making everything gay' but not for '1 Gay'. 
And when using 'Next chapter'-gesture you jump directly from 'Introduction' to 'Making everything gay'.

The Full progress bar works the other way around: it leaves out the subchapters and instead has chapter markers for "1 Gay", "Interlude", "2 Women" etc.

Is this behaviour intentional? Would it be possible to change so that it takes account of both (chapter & subchapters).
Usually this is a nonissue because the chapter and its first subchapter are on the same page. But not so with this book as there is a lot of text before the first subchapters appear. 

Reader_2019-Oct-17_140041
Reader_2019-Oct-17_140111

### PR DESCRIPTION
**Pull request by NiLuJe**
Instead of the fancyish heuristics implemented in #976 & #984, simply flatten the full list of ToC nodes, dropping duplicates in the process.

It turns out that nothing really makes use of the depth level, so, in practice, mostly everything was relying on the level 0 heuristic.

See #5500 for more details on the subject.

Fix #5500

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6716)

### CODE REVIEW

**Comment by NiLuJe** [comment]
Because I'm an idiot, I don't have "before" pictures, but here's the "after" on the testcase from https://www.mobileread.com/forums/showpost.php?p=4038495&postcount=4 (thanks!)

testcase

**Comment by NiLuJe** [comment]
I haven't actually run the testcase (which, fun fact: is the only thing actually using the depth levels directly ;p), so, heh, it might have blown up a bit ;p.

**Comment by NiLuJe** [comment]
I, err, haven't slept much this week, so, glad for reviews and tests, as I have no actual idea how I'm still conscious ^^.

**Review by Frenzie — COMMENTED** [review]
Even when my documents come with a theoretically good TOC that stuff is [still broken](
https://github.com/koreader/koreader/issues/6339) so I'm not really the person to test/comment.

**Comment by NiLuJe** [comment]
Random thought: I vaguely remember someone bench'ing `ipairs` vs? the `for i = 1, #t` approach: we don't care, right? :D.

(I vaguely tend to find ipairs more readable myself, FWIW).

**Comment by NiLuJe** [comment]
Oh, I quickly checked a random Wikipedia entry as @poire-z mentioned, and it appears to behave.

File: frontend/apps/reader/modules/readertoc.lua [hunk 1]
```diff
@@ -275,147 +280,174 @@ function ReaderToc:getTocTitleOfCurrentPage()
...
+    local ticks = self:getTocTicks()
     local ticks_candidates = {}
-    local max_level = self:getMaxDepth()
-    for i = 0, -max_level, -1 do
-        local ticks = self:getTocTicks(i)
-        table.insert(ticks_candidates, ticks)
+
+    for _, v in ipairs(ticks) do
+        for depth, page in ipairs(v) do
+            table.insert(ticks_candidates, page)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readertoc.lua:358]: You could do the duplicate/hash check here, and not insert it if page already seen. (And then, no need for the intermediate ticks_candidate.)

File: frontend/apps/reader/modules/readertoc.lua [hunk 2]
```diff
@@ -275,147 +280,174 @@ function ReaderToc:getTocTitleOfCurrentPage()
...
-        -- Find the finest toc ticks by sorting out the largest one
-        table.sort(ticks_candidates, function(a, b) return #a > #b end)
-        return ticks_candidates[1]
+
+    -- Then drop duplicates (by using the value as a key to make a quick check) (c.f., https://stackoverflow.com/a/20067270)
+    local hash = {}
+    local ticks_flattened = {}
+    for _, v in ipairs(ticks_candidates) do
+        if (not hash[v]) then
+            ticks_flattened[#ticks_flattened+1] = v
```
Comment (poire-z) [inline frontend/apps/reader/modules/readertoc.lua:367]: Or use `table.insert()` as above :)

File: frontend/apps/reader/modules/readertoc.lua [hunk 3]
```diff
@@ -275,147 +280,174 @@ function ReaderToc:getTocTitleOfCurrentPage()
...
 
-function ReaderToc:getNextChapter(cur_pageno, level)
-    local ticks = self:getTocTicks(level)
+function ReaderToc:getNextChapter(cur_pageno)
+    local ticks = self:getTocTicksFlattened()
     local next_chapter = nil
-    for i = 1, #ticks do
-        if ticks[i] > cur_pageno then
-            next_chapter = ticks[i]
+    for _, v in ipairs(ticks) do
```
Comment (poire-z) [inline frontend/apps/reader/modules/readertoc.lua:382]: (As value is a page, I would have named it p: `for _, p in ipairs(ticks) do if p > cur_pageno then` is a tad more readable/hinting.)

File: frontend/apps/reader/skimtowidget.lua [hunk 4]
```diff
@@ -359,9 +347,9 @@ end
 
 function SkimToWidget:getNextChapter(cur_pageno)
     local next_chapter = nil
-    for i = 1, #self.ticks_candidates do
-        if self.ticks_candidates[i] > cur_pageno then
-            next_chapter = self.ticks_candidates[i]
+    for _, v in ipairs(self.ticks_flattened) do
+        if v > cur_pageno then
+            next_chapter = v
```
Comment (poire-z) [inline frontend/apps/reader/skimtowidget.lua:352]: Can we re-use `ReaderToc:getNextChapter(cur_pageno)` ? `return self.ui.toc:getNextChapter(cur_pageno)` or even use that at the calling places.

**Comment by poire-z** [comment]
(Not tested yet, will do.)

So, there are really nothing using level other than 0?

Just a thought for the future, and so you can have a quick see (or even a go at it :) if that would fit in your refactoring:
People will now have many ticks on some books. We might want to have something UI to limit ticks to some level (per book, always start with all of them): so, when we have many sublevels (might not be really chapters, could be alinas, lists..., we might want to limit ticks to show only level 1 and 2 (also used by getNextChapter()) so one can browse by top-level chapters if he considers subchapters to be crap/noisy.
A little ProgressBar like for frontlight with `toc_depth` nb of squares could do :)

File: frontend/apps/reader/skimtowidget.lua (hunk 4 above)
Comment (NiLuJe) [reply frontend/apps/reader/skimtowidget.lua]: Oh, duh, didn't catch that among the rest of the code duplication in there, thanks ;).

**Comment by NiLuJe** [comment]
> (Not tested yet, will do.)
> 
> So, there are really nothing using level other than 0?
> 
> Just a thought for the future, and so you can have a quick see (or even a go at it :) if that would fit in your refactoring:
> People will now have many ticks on some books. We might want to have something UI to limit ticks to some level (per book, always start with all of them): so, when we have many sublevels (might not be really chapters, could be alinas, lists..., we might want to limit ticks to show only level 1 and 2 (also used by getNextChapter()) so one can browse by top-level chapters if he considers subchapters to be crap/noisy.
> A little ProgressBar like for frontlight with `toc_depth` nb of squares could do :)

I mean, that's certainly technically doable, I kept the full hierarchical tree intact for just such potential use-cases.

Right now, this appears to work for me, so, unless I actively find something bothering me with it, we'll see, but I probably won't dig into it further (any time soon, that is) ;).

File: frontend/apps/reader/modules/readertoc.lua [hunk 5]
```diff
@@ -275,147 +280,169 @@ function ReaderToc:getTocTitleOfCurrentPage()
...
+
+    -- It hasn't been cached yet, compute it.
+    local ticks = self:getTocTicks()
+    local ticks_flattened = {}
+
+    -- Keep track of what we add to avoid duplicates (c.f., https://stackoverflow.com/a/20067270)
+    local hash = {}
+    for _, v in ipairs(ticks) do
+        for depth, page in ipairs(v) do
+            if (not hash[page]) then
```
Comment (poire-z) [inline frontend/apps/reader/modules/readertoc.lua:360]: (Smells C :)

Comment (NiLuJe) [reply frontend/apps/reader/modules/readertoc.lua:360]: It was actually as-is in the SO snippet, and I kept it, because, indeed, I like parenthesis ;p

**Comment by poire-z** [comment]
Btw, I have in my own patch:
```lua
function ReaderToc:onPageUpdate(pageno)
    if UIManager.FULL_REFRESH_COUNT == -1 or G_reader_settings:isTrue("refresh_on_chapter_boundaries") then
        if self.avoid_chapter_flash == nil then -- avoid that on Wikipedia EPUBs
            -- No flash on chapter if too many ticks
(L71)       self.avoid_chapter_flash = (#self:getTocTicks(0) / self.ui.document:getPageCount()) > 0.3
        end
        if self.avoid_chapter_flash then
            self.pageno = pageno
            return
        end
```
which crashed with:
`frontend/apps/reader/modules/readertoc.lua:71: attempt to get length of a nil value`
So, I guess `self:getTocTicks(0)` does not work as before.
You tried to keep that working even if not used, but may be you missed the specific level=0 ?

Using getTocTicksFlattened() instead in my patch, and it works.

Checked on 2 documents, and I get the same ticks as with my:
```
-                depth = v.depth
+                -- depth = v.depth -- so we get them all and don't ignore toc entries that have children
```

So, looks good to me :) :+1:

**Comment by NiLuJe** [comment]
Oh, yeah, asking for level 0 is definitely not handled.

Basically, whatever called getTocTicks(0) should now call getTocTicksFlattened ;).

**Pull request merged**